### DISCUSSION TITLE
Api problem

### DISCUSSION THREAD

**Discussion by turancanerarkel** [emqx/emqx#11585] [General]
In the API document, when performing the Create users for authenticator operation, it asks for the Autanticator Id. what is this id? How can I find my id?

/api/v5/authentication/{id}/users

**Comment by id** [comment]
@turancanerarkel please refer to this doc: https://www.emqx.io/docs/en/v5.1/admin/api-docs.html#tag/Authentication/paths/~1authentication/post

Here is an example for built-in database authenticator and curl (this example also expects to have jq available):

```
$ TOKEN=$(curl --silent -X 'POST' "http://localhost:18083/api/v5/login" \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"username": "admin","password": "public"}' | jq -r ".token")
$ cat <<EOF >payload.json
{
  "backend": "built_in_database",
  "mechanism": "password_based",
  "password_hash_algorithm": {
    "name": "sha256",
    "salt_position": "suffix"
  },
  "user_id_type": "username"
}
EOF
$ curl -d @payload.json \
  -H "Authorization: Bearer $TOKEN" -H "Content-type: application/json" \
  http://localhost:18083/api/v5/authentication
$ curl -s \
  -H "Authorization: Bearer $TOKEN" -H "Content-type: application/json" \
  http://localhost:18083/api/v5/authentication | jq -r '.[].id'
password_based:built_in_database
$ cat <<EOF >user.json
{
  "password": "password1",
  "user_id": "user1"
}
EOF
$ curl -d @user.json \
  -H "Authorization: Bearer $TOKEN" -H "Content-type: application/json" \
  http://localhost:18083/api/v5/authentication/password_based:built_in_database/users
{"is_superuser":false,"user_id":"user1"}
```

**Comment by turancanerarkel** [comment]
But I already have two authentication installed. one mysql and one builtindb. I need to find the authenticator id rather than adding a new one. This is my real problem.

**Reply by id** [reply]
@turancanerarkel yes, I understand. Please check out the link I posted.

Also you can use this part of my snippet to list ids of available authenticators:
```
curl -s \
  -H "Authorization: Bearer $TOKEN" -H "Content-type: application/json" \
  http://localhost:18083/api/v5/authentication | jq -r '.[].id
```

**Reply by turancanerarkel** [reply]
Is the id here all or after :?

**Reply by turancanerarkel** [reply]
link : https://prnt.sc/kdqWAdj7Gio0

**Reply by id** [reply]
@turancanerarkel everything before and after `:`

**Comment by turancanerarkel** [comment]
https://prnt.sc/uFxKS5Ltdegq   

: see this link  :(

**Reply by id** [reply]
@turancanerarkel could you show which headers have you supplied with the request?

**Reply by turancanerarkel** [reply]
1: https://prnt.sc/59eApxc7Of7q

2: My authorization is Basic auth. I log in with username and password. I can get lists etc this way, but I can't create them.

**Reply by id** [reply]
@turancanerarkel could you try code from my example above to obtain auth token?

```
curl --silent -X POST "http://localhost:18083/api/v5/login" \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"username": "admin","password": "public"}'
```

I think you can do similar stuff in Postman (just need to use your own username and password).
Then you need to take the token and put it in the extra "Authorization" header: `Authorization: Bearer $TOKEN`.

EMQX API requires authentication token.

**Reply by turancanerarkel** [reply]
I get tokens and tried as you said :

1 : https://prnt.sc/UHHnY_bC67kY
2:  https://prnt.sc/51i8xX1nqnko
3 : https://prnt.sc/PAn2RWUyuJA5

**Reply by id** [reply]
Apologies, sorry for misleading you on this quest.

EMQX does not support creating users in 3rd party authentication backends via API. One has to create users there separately.

E.g. 
```
INSERT INTO mqtt_user(username, password_hash, salt, is_superuser) VALUES ('emqx_u', SHA2(concat('public', 'slat_foo123'), 256), 'slat_foo123', 1);
```

Please refer to the following documentation pages for more information:
- https://www.emqx.io/docs/en/v5.1/access-control/authn/authn.html
- https://www.emqx.io/docs/en/v5.1/access-control/authn/user_management.html
- https://www.emqx.io/docs/en/v5.1/access-control/authn/mysql.html

**Reply by turancanerarkel** [reply]
okey thank you for support
